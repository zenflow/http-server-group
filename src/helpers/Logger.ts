import { PassThrough } from 'stream'

export class Logger extends PassThrough {
  log(string: string) {
    string.split('\n').forEach(line => {
      this.write(`${line}\n`, 'utf8')
    })
  }
}
