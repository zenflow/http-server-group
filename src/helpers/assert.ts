export function assert(value: any, message: string) {
  if (!value) {
    throw new Error(`http-server-group: ${message}`)
  }
}
