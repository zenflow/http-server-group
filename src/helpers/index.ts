export * from './assert'
export * from './createProcess'
export * from './killProcessTree'
export * from './Logger'
export * from './mapStream'
export * from './rightPad'
