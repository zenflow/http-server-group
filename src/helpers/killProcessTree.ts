import treeKill from 'tree-kill'

export function killProcessTree(pid: number): Promise<void> {
  return new Promise((resolve, reject) =>
    treeKill(pid, 'SIGINT', error => (error ? reject(error) : resolve()))
  )
}
