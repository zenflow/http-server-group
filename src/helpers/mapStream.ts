import mapStreamAsync from 'map-stream'
import { Duplex } from 'stream'

export function mapStream(mapper: (arg0: string) => string): Duplex {
  return mapStreamAsync((string: string, cb: Function) =>
    cb(null, mapper(string))
  )
}
