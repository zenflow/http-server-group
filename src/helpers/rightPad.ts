export function rightPad(string: string, length: number): string {
  let result = string
  while (result.length < length) {
    result += ' '
  }
  return result
}
