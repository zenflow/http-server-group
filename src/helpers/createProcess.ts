import { spawn } from 'child_process'
import { Readable } from 'stream'
import { once } from 'events'
import mergeStream from 'merge-stream'
import splitStream from 'split'
import { mapStream } from './mapStream'
import { killProcessTree } from './killProcessTree'

export function createProcess(command: string | Array<string>, env: object) {
  const proc = Array.isArray(command)
    ? spawn(command[0], command.slice(1), {
        env: { ...process.env, ...env },
      })
    : spawn(command, {
        shell: true,
        env: { ...process.env, ...env },
      })
  const outputStream = (mergeStream(
    proc.stdout
      .setEncoding('utf8')
      .pipe(splitStream())
      .pipe(mapStream(line => `[out] ${line}\n`)),
    proc.stderr
      .setEncoding('utf8')
      .pipe(splitStream())
      .pipe(mapStream(line => `[ERR] ${line}\n`))
  ) as unknown) as Readable
  let didExit = false
  const exited = once(outputStream, 'end').then(() => {
    didExit = true
  })
  let killPromise: Promise<void> | null = null
  function kill(): Promise<void> {
    if (!killPromise) {
      killPromise = Promise.resolve().then(async () => {
        if (!didExit) {
          await killProcessTree(proc.pid)
        }
        await exited
      })
    }
    return killPromise
  }
  return { outputStream, exited, kill }
}
