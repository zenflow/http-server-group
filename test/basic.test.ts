// @ts-ignore
import { fetchText } from './helpers/fetchText'
// @ts-ignore
import { createServerGroupProcess } from './helpers/createServerGroupProcess'

const port = 3000

describe('basic', () => {
  it('works', async () => {
    const proc = await createServerGroupProcess(port, {
      servers: [
        {
          label: 'a',
          command: `node ${__dirname}/fixtures/server-a.js`,
          port: 3001,
          paths: ['/a'],
        },
        {
          label: 'b',
          command: `node ${__dirname}/fixtures/server-b.js`,
          port: 3002,
          paths: ['/b'],
        },
      ],
    })
    const [aText, bText] = await Promise.all([
      fetchText(`http://localhost:${port}/a`),
      fetchText(`http://localhost:${port}/b`),
    ])
    expect(aText).toEqual('a')
    expect(bText).toEqual('b')
    await proc.destroy()
  })
})
