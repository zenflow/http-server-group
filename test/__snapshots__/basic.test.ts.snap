// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic has consistent output 1`] = `
Object {
  "": Array [],
  "$manager": Array [
    "$manager | http-server-group config {",
    "$manager |   \\"proxyOptions\\": {},",
    "$manager |   \\"servers\\": [",
    "$manager |     {",
    "$manager |       \\"label\\": \\"a\\",",
    "$manager |       \\"env\\": {",
    "$manager |         \\"KEY\\": \\"a\\"",
    "$manager |       },",
    "$manager |       \\"command\\": [",
    "$manager |         \\"node\\",",
    "$manager |         \\"test/fixtures/server-node.js\\"",
    "$manager |       ],",
    "$manager |       \\"host\\": \\"localhost\\",",
    "$manager |       \\"port\\": 3001,",
    "$manager |       \\"paths\\": [",
    "$manager |         \\"/a\\"",
    "$manager |       ],",
    "$manager |       \\"proxyOptions\\": {}",
    "$manager |     },",
    "$manager |     {",
    "$manager |       \\"label\\": \\"b\\",",
    "$manager |       \\"env\\": {",
    "$manager |         \\"KEY\\": \\"b\\"",
    "$manager |       },",
    "$manager |       \\"command\\": \\"node test/fixtures/server-node.js\\",",
    "$manager |       \\"host\\": \\"localhost\\",",
    "$manager |       \\"port\\": 3002,",
    "$manager |       \\"paths\\": [",
    "$manager |         \\"/b\\"",
    "$manager |       ],",
    "$manager |       \\"proxyOptions\\": {}",
    "$manager |     }",
    "$manager |   ]",
    "$manager | }",
    "$manager | Starting 'a'...",
    "$manager | Starting 'b'...",
    "$manager | Started 'a' @ http://localhost:3001",
    "$manager | Started 'b' @ http://localhost:3002",
    "$manager | Starting '$proxy'...",
    "$manager | Started '$proxy' @ http://localhost:3000",
    "$manager | Ready",
  ],
  "$proxy": Array [],
  "a": Array [
    "a        | [out] Starting...",
    "a        | [out] Started",
  ],
  "b": Array [
    "b        | [out] Starting...",
    "b        | [out] Started",
  ],
}
`;

exports[`basic has consistent output 2`] = `
Array [
  "",
  "",
]
`;
